File: app/src/main/java/tr/edu/iyte/quickreply/adapters/ReplyAdapter.java
package tr.edu.iyte.quickreply.adapters;

import android.graphics.Color;
import android.support.v4.view.MotionEventCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import tr.edu.iyte.quickreply.QuickReplyTile;
import tr.edu.iyte.quickreply.R;
import tr.edu.iyte.quickreply.interfaces.ReplyItemTouchHelperAdapter;
import tr.edu.iyte.quickreply.interfaces.ViewHolderInteractionListener;
import tr.edu.iyte.quickreply.interfaces.OnReplyInteractedListener;
import tr.edu.iyte.quickreply.interfaces.OnStartDragListener;

public class ReplyAdapter
        extends RecyclerView.Adapter<ReplyAdapter.ViewHolder>
        implements ReplyItemTouchHelperAdapter {

    private final List<String> replies;
    private final OnStartDragListener sdListener;
    private final OnReplyInteractedListener riListener;

    public ReplyAdapter(List<String> replies, OnStartDragListener sdListener, OnReplyInteractedListener riListener) {
        super();
        this.replies = replies;
        this.sdListener = sdListener;
        this.riListener = riListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.reply, parent, false));
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.textView.setText(replies.get(position));

        holder.itemView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if(MotionEventCompat.getActionMasked(event) == MotionEvent.ACTION_DOWN) {
                    sdListener.onStartDrag(holder);
                }
                return false;
            }
        });

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                riListener.onReplySelected(holder.textView.getText().toString());
            }
        });
    }

    @Override
    public void onReplyDismissed(int position) {
        remove(position);
        riListener.onReplyDismissed();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return replies.size();
    }

    public void add(String reply) {
        replies.add(reply);
        notifyItemInserted(replies.size() - 1);
    }

    private void remove(int idx) {
        QuickReplyTile.removeReply(replies.get(idx));
        replies.remove(idx);
        notifyItemRemoved(idx);
    }

    public static class ViewHolder
            extends RecyclerView.ViewHolder
            implements ViewHolderInteractionListener {
        private TextView textView;

        private ViewHolder(View v) {
            super(v);
            textView = (TextView) v.findViewById(R.id.reply_text);
        }

        @Override
        public void onInteractionStart() {
            itemView.setBackgroundColor(Color.LTGRAY);
        }

        @Override
        public void onInteractionEnd() {
            itemView.setBackgroundColor(0);
        }
    }
}

File: app/src/main/java/tr/edu/iyte/quickreply/adapters/ReplyAdapter.kt
package tr.edu.iyte.quickreply.adapters

import android.graphics.Color
import android.support.v4.view.MotionEventCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import tr.edu.iyte.quickreply.interfaces.ReplyItemTouchHelperAdapter
import org.jetbrains.anko.find
import tr.edu.iyte.quickreply.R
import tr.edu.iyte.quickreply.ReplyManager
import tr.edu.iyte.quickreply.helper.asSameAs
import tr.edu.iyte.quickreply.interfaces.OnReplyInteractedListener
import tr.edu.iyte.quickreply.interfaces.OnStartDragListener
import tr.edu.iyte.quickreply.interfaces.ViewHolderInteractionListener

class ReplyAdapter(private val replies: MutableList<String>,
                   private val startDragListener: OnStartDragListener,
                   private val replyInteractedListener: OnReplyInteractedListener) :
        RecyclerView.Adapter<ReplyAdapter.ViewHolder>(),
        ReplyItemTouchHelperAdapter {
    class ViewHolder(v: View) :
            RecyclerView.ViewHolder(v),
            ViewHolderInteractionListener {
        val textView: TextView = v.find(R.id.reply_text)

        override fun onInteractionStart() = itemView.setBackgroundColor(Color.LTGRAY)
        override fun onInteractionEnd() = itemView.setBackgroundColor(0)
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int)
            = ViewHolder(LayoutInflater
                .from(parent!!.context)
                .inflate(R.layout.list_item_reply, parent, false))

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        holder?.textView!!.text = replies[position]

        holder.itemView.setOnTouchListener { _, event ->
            if(MotionEventCompat.getActionMasked(event) == MotionEvent.ACTION_DOWN)
                startDragListener.onStartDrag(holder)
            false
        }

        holder.itemView.setOnClickListener {
            replyInteractedListener.onReplySelected(holder.textView.text.toString())
        }
    }

    override fun onReplyDismissed(position: Int) {
        remove(replies[position])
        replyInteractedListener.onReplyDismissed()
    }

    fun add(reply: String) {
        replies.add(reply)
        notifyItemInserted(replies.size - 1)
    }

    fun addAll(replies: Collection<String>) {
        val rangeStart = this.replies.size
        this.replies.addAll(replies)
        notifyItemRangeInserted(rangeStart, replies.size)
    }

    fun updateIfModified() {
        if(!replies.asSameAs(ReplyManager.replies)) {
            replies.clear()
            replies.addAll(ReplyManager.replies)
            notifyDataSetChanged()
        }
    }

    private fun remove(reply: String) {
        ReplyManager.removeReply(reply)
        val i = replies.indexOf(reply)
        replies.removeAt(i)
        notifyItemRemoved(i)
    }

    override fun getItemId(position: Int) = position.toLong()
    override fun getItemCount() = replies.size
}